File: packages/guards/src/code-block-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface CodeBlockSafetyOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /rm\s+-rf\s+\//,
  /DROP\s+TABLE/i,
  /DROP\s+DATABASE/i,
  /eval\s*\(\s*(user|input|req|request)/i,
  /password\s*=\s*["'][^"']{3,}["']/i,
  /secret\s*=\s*["'][^"']{3,}["']/i,
  /api_?key\s*=\s*["'][^"']{3,}["']/i,
  /exec\s*\(\s*["'].*\bsh\b/i,
  /os\.system\s*\(/i,
  /subprocess\.call\s*\(\s*["'].*\brm\b/i,
  /chmod\s+777/,
  /mkfs\s+/,
  />\s*\/dev\/sda/,
  /format\s+[cC]:/i,
];

export function codeBlockSafety(options: CodeBlockSafetyOptions): Guard {
  return {
    name: 'code-block-safety',
    version: '0.1.0',
    description: 'Safety checks for code blocks in AI responses',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'code-block-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/code_block_safety.py
"""Safety checks for code blocks in AI responses."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"rm\s+-rf\s+/"),
    re.compile(r"DROP\s+TABLE", re.IGNORECASE),
    re.compile(r"DROP\s+DATABASE", re.IGNORECASE),
    re.compile(
        r"eval\s*\(\s*(user|input|req|request)",
        re.IGNORECASE,
    ),
    re.compile(
        r"password\s*=\s*[\"'][^\"']{3,}[\"']",
        re.IGNORECASE,
    ),
    re.compile(
        r"secret\s*=\s*[\"'][^\"']{3,}[\"']",
        re.IGNORECASE,
    ),
    re.compile(
        r"api_?key\s*=\s*[\"'][^\"']{3,}[\"']",
        re.IGNORECASE,
    ),
    re.compile(
        r"exec\s*\(\s*[\"'].*\bsh\b", re.IGNORECASE
    ),
    re.compile(r"os\.system\s*\(", re.IGNORECASE),
    re.compile(
        r"subprocess\.call\s*\(\s*[\"'].*\brm\b",
        re.IGNORECASE,
    ),
    re.compile(r"chmod\s+777"),
    re.compile(r"mkfs\s+"),
    re.compile(r">\s*/dev/sda"),
    re.compile(r"format\s+[cC]:", re.IGNORECASE),
]


class _CodeBlockSafety:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "code-block-safety"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in _PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="code-block-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Unsafe code pattern detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def code_block_safety(
    *, action: str = "block"
) -> _CodeBlockSafety:
    return _CodeBlockSafety(action=action)
